File: backend/src/config/serviceCatalog.js
/**
 * The five services Khatu Pay offers, and the rules that apply to them.
 *
 * Why this exists: ClubAPI's operatorList returns only `name`, `operatorId`
 * and `gstMode`. It does not return BBPS biller IDs, nor the input fields a
 * biller needs ("Consumer number", "Last 4 digits of card"...). So which
 * billers exist, their bbpsId and their input fields are stored per provider
 * (see models/ServiceProvider.js, managed from the admin panel), and this file
 * holds the defaults and validation shared by the API, the payment flow and
 * the admin sync.
 *
 * Field `key` is the ClubAPI request parameter the value is sent as:
 *   mobile   -> the primary account value ("mobile number or consumer number")
 *   opvalue1..opvalue5 -> additional values some billers require
 *
 * Everything here is pure (no I/O) so it can be unit tested.
 */

export const SERVICE_KEYS = ['mobile', 'dth', 'credit_card', 'electricity', 'fastag'];
export const FIELD_KEYS = ['mobile', 'opvalue1', 'opvalue2', 'opvalue3', 'opvalue4', 'opvalue5'];
export const INPUT_TYPES = ['mobile', 'number', 'text'];

/** Maximum Razorpay order this app creates (mirrors routes/payments.js). */
export const MAX_ORDER_AMOUNT = 100000;

/** A fetched bill can be paid for this long before it must be fetched again. */
export const BILL_FETCH_TTL_MS = 30 * 60 * 1000;

const mobileField = (label = 'Mobile number') => ({
  key: 'mobile',
  label,
  placeholder: '10-digit mobile number',
  hint: '',
  inputType: 'mobile',
  minLength: 10,
  maxLength: 10,
  pattern: '^[6-9][0-9]{9}$',
  uppercase: false,
});

export const SERVICE_DEFINITIONS = {
  mobile: {
    key: 'mobile',
    title: 'Mobile Recharge',
    subtitle: 'Prepaid recharge for all operators',
    kind: 'recharge',
    paymentType: 'RECHARGE',
    providerLabel: 'Operator',
    minAmount: 10,
    maxAmount: 10000,
    wholeRupees: true,
    supportsPlans: true,
    supportsOperatorDetect: true,
    defaultFields: [mobileField('Mobile number')],
    settingsFlag: 'mobileRechargeEnabled',
  },
  dth: {
    key: 'dth',
    title: 'DTH Recharge',
    subtitle: 'Tata Play, Airtel, Dish TV & more',
    kind: 'recharge',
    paymentType: 'RECHARGE',
    providerLabel: 'DTH operator',
    minAmount: 100,
    maxAmount: 25000,
    wholeRupees: true,
    supportsPlans: false,
    supportsOperatorDetect: false,
    defaultFields: [
      {
        key: 'mobile',
        label: 'Subscriber ID / Customer ID',
        placeholder: 'Registered mobile or subscriber ID',
        hint: 'Shown on your set-top box welcome screen or recharge SMS',
        inputType: 'number',
        minLength: 6,
        maxLength: 16,
        pattern: '^[0-9]{6,16}$',
        uppercase: false,
      },
    ],
    settingsFlag: 'dthRechargeEnabled',
  },
  credit_card: {
    key: 'credit_card',
    title: 'Credit Card Bill',
    subtitle: 'Pay any bank credit card via BBPS',
    kind: 'bill',
    paymentType: 'BBPS_BILL',
    providerLabel: 'Card issuing bank',
    minAmount: 1,
    maxAmount: MAX_ORDER_AMOUNT,
    wholeRupees: false,
    supportsPlans: false,
    supportsOperatorDetect: false,
    defaultFields: [
      mobileField('Registered mobile number'),
      {
        key: 'opvalue1',
        label: 'Last 4 digits of credit card',
        placeholder: 'XXXX',
        hint: 'Only the last 4 digits. Never share your full card number or CVV.',
        inputType: 'number',
        minLength: 4,
        maxLength: 4,
        pattern: '^[0-9]{4}$',
        uppercase: false,
      },
    ],
    settingsFlag: 'billPaymentEnabled',
  },
  electricity: {
    key: 'electricity',
    title: 'Electricity Bill',
    subtitle: 'All state electricity boards',
    kind: 'bill',
    paymentType: 'BBPS_BILL',
    providerLabel: 'Electricity board',
    minAmount: 1,
    maxAmount: MAX_ORDER_AMOUNT,
    wholeRupees: false,
    supportsPlans: false,
    supportsOperatorDetect: false,
    defaultFields: [
      {
        key: 'mobile',
        label: 'Consumer number',
        placeholder: 'As printed on your electricity bill',
        hint: 'Also called CA number, account number or K number',
        inputType: 'text',
        minLength: 4,
        maxLength: 24,
        pattern: '^[A-Z0-9/-]{4,24}$',
        uppercase: true,
      },
    ],
    settingsFlag: 'billPaymentEnabled',
  },
  fastag: {
    key: 'fastag',
    title: 'FASTag Recharge',
    subtitle: 'Recharge FASTag of any issuer bank',
    kind: 'bill',
    paymentType: 'BBPS_BILL',
    providerLabel: 'FASTag issuer bank',
    minAmount: 100,
    maxAmount: MAX_ORDER_AMOUNT,
    wholeRupees: true,
    supportsPlans: false,
    supportsOperatorDetect: false,
    defaultFields: [
      {
        key: 'mobile',
        label: 'Vehicle registration number',
        placeholder: 'e.g. MH12AB1234',
        hint: 'Enter without spaces',
        inputType: 'text',
        minLength: 6,
        maxLength: 12,
        pattern: '^[A-Z0-9]{6,12}$',
        uppercase: true,
      },
    ],
    settingsFlag: 'billPaymentEnabled',
  },
};

/**
 * Providers whose IDs are published in ClubAPI's own documentation, used to
 * seed an empty catalog. Everything else is added through "Sync from ClubAPI"
 * or entered by an admin from the ClubAPI panel's biller list - biller IDs are
 * never guessed.
 */
export const DOCUMENTED_PROVIDERS = [
  { service: 'mobile', name: 'Airtel', operatorId: '1', sortOrder: 1 },
  { service: 'mobile', name: 'Vi (Vodafone Idea)', operatorId: '2', sortOrder: 2 },
  { service: 'mobile', name: 'Jio', operatorId: '3', sortOrder: 3 },
  { service: 'mobile', name: 'BSNL', operatorId: '4', sortOrder: 4 },
  {
    service: 'electricity',
    name: 'Uttar Pradesh Power Corp Ltd (UPPCL) - Urban',
    bbpsId: 'UPPCL0000UTP02',
    state: 'Uttar Pradesh',
    sortOrder: 1,
    fields: [
      {
        key: 'mobile',
        label: 'Consumer / account number',
        placeholder: '10 to 12 digit account number',
        hint: 'Printed as "Account No." on your UPPCL bill',
        inputType: 'number',
        minLength: 10,
        maxLength: 12,
        pattern: '^[0-9]{10,12}$',
        uppercase: false,
      },
    ],
  },
];

/* ----------------------------------------------------------------- fields */

/**
 * Sanitises an admin-supplied field list. Invalid entries are dropped rather
 * than rejected so a partially wrong edit never takes a biller offline.
 */
export function sanitizeFields(fields) {
  if (!Array.isArray(fields)) return [];
  const seen = new Set();
  const result = [];

  for (const raw of fields) {
    if (!raw || typeof raw !== 'object') continue;
    const key = String(raw.key || '').trim();
    const label = String(raw.label || '').trim();
    if (!FIELD_KEYS.includes(key) || !label || seen.has(key)) continue;

    let pattern = String(raw.pattern || '').trim();
    if (pattern) {
      try {
        new RegExp(pattern);
      } catch {
        pattern = '';
      }
    }

    const minLength = Math.max(0, Math.min(64, parseInt(raw.minLength, 10) || 0));
    const maxLength = Math.max(minLength || 1, Math.min(64, parseInt(raw.maxLength, 10) || 64));

    seen.add(key);
    result.push({
      key,
      label: label.slice(0, 60),
      placeholder: String(raw.placeholder || '').trim().slice(0, 80),
      hint: String(raw.hint || '').trim().slice(0, 160),
      inputType: INPUT_TYPES.includes(raw.inputType) ? raw.inputType : 'text',
      minLength,
      maxLength,
      pattern,
      uppercase: raw.uppercase === true,
    });
  }

  // The primary `mobile` parameter is mandatory for every ClubAPI call, so it
  // always sorts first.
  return result.sort((a, b) => FIELD_KEYS.indexOf(a.key) - FIELD_KEYS.indexOf(b.key));
}

/** The fields a provider actually uses: its own, else the service defaults. */
export function effectiveFields(serviceKey, providerFields) {
  const own = sanitizeFields(providerFields);
  if (own.some((field) => field.key === 'mobile')) return own;
  return sanitizeFields(SERVICE_DEFINITIONS[serviceKey]?.defaultFields || []);
}

/**
 * Validates and normalises user input against field definitions.
 * @returns {{ values: Record<string,string>, errors: Record<string,string> }}
 */
export function validateFieldValues(fields, input) {
  const values = {};
  const errors = {};
  const source = input && typeof input === 'object' ? input : {};

  for (const field of fields) {
    let value = String(source[field.key] ?? '').trim();
    if (field.inputType === 'mobile' || field.inputType === 'number') {
      value = value.replace(/[\s-]/g, '');
    } else {
      value = value.replace(/\s+/g, '');
    }
    if (field.uppercase) value = value.toUpperCase();

    if (!value) {
      errors[field.key] = `Enter ${field.label.toLowerCase()}`;
      continue;
    }
    if ((field.inputType === 'mobile' || field.inputType === 'number') && !/^\d+$/.test(value)) {
      errors[field.key] = `${field.label} should contain digits only`;
      continue;
    }
    if (field.minLength && value.length < field.minLength) {
      errors[field.key] =
        field.minLength === field.maxLength
          ? `${field.label} must be ${field.minLength} characters`
          : `${field.label} must be at least ${field.minLength} characters`;
      continue;
    }
    if (field.maxLength && value.length > field.maxLength) {
      errors[field.key] = `${field.label} can be at most ${field.maxLength} characters`;
      continue;
    }
    if (field.pattern && !new RegExp(field.pattern).test(value)) {
      errors[field.key] = `Enter a valid ${field.label.toLowerCase()}`;
      continue;
    }
    values[field.key] = value;
  }

  return { values, errors };
}

/* ----------------------------------------------------------------- amounts */

/**
 * Validates a payable amount for a service, optionally narrowed by a fetched
 * bill's exactness rule. Returns the amount rounded to paise, or an error.
 */
export function validateServiceAmount(serviceKey, rawAmount, rule = null) {
  const definition = SERVICE_DEFINITIONS[serviceKey];
  if (!definition) return { error: 'Unknown service' };

  const amount = Math.round(Number(rawAmount) * 100) / 100;
  if (!Number.isFinite(amount) || amount <= 0) return { error: 'Enter a valid amount' };
  if (definition.wholeRupees && !Number.isInteger(amount)) {
    return { error: 'Amount must be in whole rupees' };
  }

  const min = Math.max(definition.minAmount, rule?.min ?? definition.minAmount);
  const max = Math.min(definition.maxAmount, rule?.max ?? definition.maxAmount);

  if (rule && rule.editable === false && Math.abs(amount - rule.min) > 0.009) {
    return { error: `This biller accepts only the exact bill amount of ₹${rule.min}` };
  }
  if (amount < min) return { error: `Minimum amount is ₹${min}` };
  if (amount > max) return { error: `Maximum amount is ₹${max}` };

  return { amount };
}

/* ------------------------------------------------------------ classification */

const DTH_PATTERN = /\b(dth|digital\s*tv|dish\s*tv|dishtv|tata\s*(play|sky)|sun\s*direct|d2h|videocon)\b/i;
const MOBILE_PATTERN = /\b(airtel|vodafone|idea|vi|jio|bsnl|mtnl)\b/i;
const NOT_PREPAID_PATTERN =
  /(postpaid|post\s*paid|landline|broadband|fiber|fibre|fastag|money|wallet|payments?\s*bank|credit|insurance|loan|\bgas\b|water|electric|\bdth\b|\btv\b)/i;
const CREDIT_CARD_PATTERN = /credit\s*card|\bcards?\b/i;
// "Fastag", "FASTag", "Fast Tag" and "FastTag" all occur in biller names.
const FASTAG_PATTERN = /fas\s*t?\s*tag|\bnetc\b/i;
const ELECTRICITY_PATTERN =
  /(electric|electricity|power|vidyut|bijli|energy|discom|bses|msedcl|mahavitaran|uppcl|tneb|tangedco|kseb|bescom|mescom|hescom|gescom|cesc|torrent|tata\s*power|adani\s*electricity|jvvnl|avvnl|jdvvnl|pspcl|dhbvn|uhbvn|wbsedcl|apepdcl|apspdcl|tsspdcl|tsnpdcl|mppkvvcl|cspdcl|nbpdcl|sbpdcl|jbvnl|upcl|hpseb|dgvcl|mgvcl|pgvcl|ugvcl|noida\s*power|apdcl|tsecl|mspdcl|nesco|wesco|southco|tpcodl|tpsodl|tpwodl|tpnodl|best\s*undertaking|kedl)/i;
const NOT_ELECTRICITY_PATTERN = /(gas|water|insurance|loan|broadband|fastag|card|lpg|cylinder)/i;

/**
 * Decides which of the five services a ClubAPI operator belongs to, from its
 * name and (when ClubAPI sends one) its category. Returns null for anything
 * outside the five services.
 */
export function classifyOperator({ name = '', category = '', bbpsId = '' } = {}) {
  const text = `${category} ${name}`.trim();
  const hasBbps = Boolean(String(bbpsId || '').trim());

  if (hasBbps) {
    if (FASTAG_PATTERN.test(text)) return 'fastag';
    if (CREDIT_CARD_PATTERN.test(text) && !/loan|insurance/i.test(text)) return 'credit_card';
    if (ELECTRICITY_PATTERN.test(text) && !NOT_ELECTRICITY_PATTERN.test(text)) return 'electricity';
    return null;
  }

  if (DTH_PATTERN.test(text)) return 'dth';
  if (MOBILE_PATTERN.test(name) && !NOT_PREPAID_PATTERN.test(text)) return 'mobile';
  return null;
}

/** Public shape of a service for the app. */
export function publicServiceDefinition(definition) {
  return {
    key: definition.key,
    title: definition.title,
    subtitle: definition.subtitle,
    kind: definition.kind,
    providerLabel: definition.providerLabel,
    minAmount: definition.minAmount,
    maxAmount: definition.maxAmount,
    wholeRupees: definition.wholeRupees,
    supportsPlans: definition.supportsPlans,
    supportsOperatorDetect: definition.supportsOperatorDetect,
  };
}

File: khatupay_app/lib/clubapi/models/service_catalog.dart
import 'package:flutter/services.dart';

/// Models for the five recharge & bill services (Mobile, DTH, Credit Card,
/// Electricity, FASTag), mirroring `/api/services/*`.
///
/// Every parser is null-tolerant: a missing or oddly typed field falls back to
/// a safe default instead of throwing, so one bad row from the API can never
/// crash a payment screen.

String _text(dynamic value) => value?.toString().trim() ?? '';

double _double(dynamic value) {
  if (value is num) return value.toDouble();
  return double.tryParse(_text(value).replaceAll(',', '')) ?? 0;
}

int _int(dynamic value, [int fallback = 0]) {
  if (value is int) return value;
  if (value is num) return value.toInt();
  return int.tryParse(_text(value)) ?? fallback;
}

bool _bool(dynamic value, [bool fallback = false]) {
  if (value is bool) return value;
  final text = _text(value).toLowerCase();
  if (text == 'true') return true;
  if (text == 'false') return false;
  return fallback;
}

Map<String, dynamic> _map(dynamic value) =>
    value is Map ? Map<String, dynamic>.from(value) : <String, dynamic>{};

List<Map<String, dynamic>> _maps(dynamic value) => value is List
    ? value.whereType<Map>().map((row) => Map<String, dynamic>.from(row)).toList()
    : const [];

DateTime? _date(dynamic value) {
  final text = _text(value);
  return text.isEmpty ? null : DateTime.tryParse(text)?.toLocal();
}

/// One input a provider needs (e.g. "Consumer number"), sent to ClubAPI as
/// [key] (`mobile` or `opvalue1`..`opvalue5`).
class ServiceField {
  const ServiceField({
    required this.key,
    required this.label,
    this.placeholder = '',
    this.hint = '',
    this.inputType = 'text',
    this.minLength = 0,
    this.maxLength = 64,
    this.pattern = '',
    this.uppercase = false,
  });

  factory ServiceField.fromJson(Map<String, dynamic> json) => ServiceField(
        key: _text(json['key']).isEmpty ? 'mobile' : _text(json['key']),
        label: _text(json['label']).isEmpty ? 'Account number' : _text(json['label']),
        placeholder: _text(json['placeholder']),
        hint: _text(json['hint']),
        inputType: _text(json['inputType']).isEmpty ? 'text' : _text(json['inputType']),
        minLength: _int(json['minLength']),
        maxLength: _int(json['maxLength'], 64).clamp(1, 64),
        pattern: _text(json['pattern']),
        uppercase: _bool(json['uppercase']),
      );

  final String key;
  final String label;
  final String placeholder;
  final String hint;
  final String inputType;
  final int minLength;
  final int maxLength;
  final String pattern;
  final bool uppercase;

  bool get isNumeric => inputType == 'mobile' || inputType == 'number';

  TextInputType get keyboardType => isNumeric ? TextInputType.number : TextInputType.text;

  List<TextInputFormatter> get formatters => [
        if (isNumeric) FilteringTextInputFormatter.digitsOnly,
        if (!isNumeric) FilteringTextInputFormatter.deny(RegExp(r'\s')),
        LengthLimitingTextInputFormatter(maxLength),
        if (uppercase) const _UpperCaseFormatter(),
      ];

  /// Mirrors the server's normalisation so client and server agree on what
  /// "valid" means.
  String normalize(String raw) {
    var value = raw.trim();
    value = isNumeric ? value.replaceAll(RegExp(r'[\s-]'), '') : value.replaceAll(RegExp(r'\s+'), '');
    return uppercase ? value.toUpperCase() : value;
  }

  /// Returns an error message, or null when [raw] is valid.
  String? validate(String raw) {
    final value = normalize(raw);
    if (value.isEmpty) return 'Enter ${label.toLowerCase()}';
    if (isNumeric && !RegExp(r'^\d+$').hasMatch(value)) {
      return '$label should contain digits only';
    }
    if (minLength > 0 && value.length < minLength) {
      return minLength == maxLength
          ? '$label must be $minLength characters'
          : '$label must be at least $minLength characters';
    }
    if (value.length > maxLength) return '$label can be at most $maxLength characters';
    if (pattern.isNotEmpty) {
      try {
        if (!RegExp(pattern).hasMatch(value)) return 'Enter a valid ${label.toLowerCase()}';
      } catch (_) {
        // An invalid admin pattern must never block the customer.
      }
    }
    return null;
  }
}

class _UpperCaseFormatter extends TextInputFormatter {
  const _UpperCaseFormatter();

  @override
  TextEditingValue formatEditUpdate(TextEditingValue oldValue, TextEditingValue newValue) =>
      newValue.copyWith(text: newValue.text.toUpperCase());
}

/// An operator (Mobile / DTH) or a BBPS biller (Credit Card / Electricity /
/// FASTag).
class ServiceProviderItem {
  const ServiceProviderItem({
    required this.id,
    required this.service,
    required this.name,
    this.state = '',
    this.fields = const [],
    this.supportsFetch = false,
  });

  factory ServiceProviderItem.fromJson(Map<String, dynamic> json) => ServiceProviderItem(
        id: _text(json['id']),
        service: _text(json['service']),
        name: _text(json['name']).isEmpty ? 'Provider' : _text(json['name']),
        state: _text(json['state']),
        fields: _maps(json['fields']).map(ServiceField.fromJson).toList(),
        supportsFetch: _bool(json['supportsFetch']),
      );

  final String id;
  final String service;
  final String name;
  final String state;
  final List<ServiceField> fields;
  final bool supportsFetch;

  /// Two-letter monogram for the provider avatar.
  String get initials {
    final words = name
        .replaceAll(RegExp(r'[^A-Za-z0-9 ]'), ' ')
        .split(RegExp(r'\s+'))
        .where((word) => word.isNotEmpty)
        .toList();
    if (words.isEmpty) return 'KP';
    if (words.length == 1) {
      final word = words.first;
      return (word.length >= 2 ? word.substring(0, 2) : word).toUpperCase();
    }
    return (words[0][0] + words[1][0]).toUpperCase();
  }
}

/// One of the five services.
class BillServiceItem {
  const BillServiceItem({
    required this.key,
    required this.title,
    this.subtitle = '',
    this.kind = 'recharge',
    this.providerLabel = 'Operator',
    this.minAmount = 1,
    this.maxAmount = 100000,
    this.wholeRupees = true,
    this.supportsPlans = false,
    this.supportsOperatorDetect = false,
    this.available = true,
    this.unavailableReason,
    this.defaultFields = const [],
    this.providers = const [],
  });

  factory BillServiceItem.fromJson(Map<String, dynamic> json) => BillServiceItem(
        key: _text(json['key']),
        title: _text(json['title']),
        subtitle: _text(json['subtitle']),
        kind: _text(json['kind']).isEmpty ? 'recharge' : _text(json['kind']),
        providerLabel: _text(json['providerLabel']).isEmpty ? 'Operator' : _text(json['providerLabel']),
        minAmount: _double(json['minAmount']) > 0 ? _double(json['minAmount']) : 1,
        maxAmount: _double(json['maxAmount']) > 0 ? _double(json['maxAmount']) : 100000,
        wholeRupees: _bool(json['wholeRupees'], true),
        supportsPlans: _bool(json['supportsPlans']),
        supportsOperatorDetect: _bool(json['supportsOperatorDetect']),
        available: _bool(json['available'], true),
        unavailableReason: _text(json['unavailableReason']).isEmpty ? null : _text(json['unavailableReason']),
        defaultFields: _maps(json['defaultFields']).map(ServiceField.fromJson).toList(),
        providers: _maps(json['providers']).map(ServiceProviderItem.fromJson).where((p) => p.id.isNotEmpty).toList(),
      );

  final String key;
  final String title;
  final String subtitle;
  final String kind;
  final String providerLabel;
  final double minAmount;
  final double maxAmount;
  final bool wholeRupees;
  final bool supportsPlans;
  final bool supportsOperatorDetect;
  final bool available;
  final String? unavailableReason;
  final List<ServiceField> defaultFields;
  final List<ServiceProviderItem> providers;

  bool get isRecharge => kind == 'recharge';
  bool get isBill => kind == 'bill';

  /// Account field for a recharge provider (falls back to the service default).
  ServiceField accountFieldFor(ServiceProviderItem? provider) {
    final fields = provider?.fields ?? const <ServiceField>[];
    for (final field in fields) {
      if (field.key == 'mobile') return field;
    }
    for (final field in defaultFields) {
      if (field.key == 'mobile') return field;
    }
    return const ServiceField(key: 'mobile', label: 'Account number');
  }

  /// Validates an amount for this service (and optionally a fetched bill's
  /// rule). Returns an error message or null.
  String? validateAmount(double? amount, {AmountRule? rule}) {
    if (amount == null || amount <= 0) return 'Enter an amount';
    if (wholeRupees && amount != amount.roundToDouble()) return 'Enter amount in whole rupees';
    final min = rule == null ? minAmount : (rule.min > minAmount ? rule.min : minAmount);
    final max = rule == null ? maxAmount : (rule.max < maxAmount ? rule.max : maxAmount);
    if (rule != null && !rule.editable && (amount - rule.min).abs() > 0.009) {
      return 'This biller accepts only the exact bill amount';
    }
    if (amount < min) return 'Minimum amount is ₹${_plain(min)}';
    if (amount > max) return 'Maximum amount is ₹${_plain(max)}';
    return null;
  }
}

String _plain(double value) =>
    value == value.roundToDouble() ? value.toStringAsFixed(0) : value.toStringAsFixed(2);

class ServiceCatalog {
  const ServiceCatalog(this.services);

  factory ServiceCatalog.fromJson(Map<String, dynamic> json) => ServiceCatalog(
        _maps(json['services']).map(BillServiceItem.fromJson).where((s) => s.key.isNotEmpty).toList(),
      );

  final List<BillServiceItem> services;

  BillServiceItem? byKey(String key) {
    for (final service in services) {
      if (service.key == key) return service;
    }
    return null;
  }
}

/// How much may be paid against a fetched bill (from the biller's exactness).
class AmountRule {
  const AmountRule({
    this.mode = 'ANY',
    this.editable = true,
    this.min = 1,
    this.max = 100000,
    this.suggested,
  });

  factory AmountRule.fromJson(Map<String, dynamic> json) => AmountRule(
        mode: _text(json['mode']).isEmpty ? 'ANY' : _text(json['mode']),
        editable: _bool(json['editable'], true),
        min: _double(json['min']),
        max: _double(json['max']) > 0 ? _double(json['max']) : 100000,
        suggested: json['suggested'] == null ? null : _double(json['suggested']),
      );

  final String mode;
  final bool editable;
  final double min;
  final double max;
  final double? suggested;
}

/// A bill fetched and stored by the server. [fetchId] is what payment uses.
class FetchedBill {
  const FetchedBill({
    required this.fetchId,
    required this.service,
    required this.providerId,
    required this.providerName,
    required this.accountRef,
    required this.rule,
    this.customerName = '',
    this.amount = 0,
    this.dueDate = '',
    this.billDate = '',
    this.billNumber = '',
    this.billPeriod = '',
    this.billerName = '',
    this.billerBalance = '',
    this.remarks = '',
    this.expiresAt,
  });

  factory FetchedBill.fromJson(Map<String, dynamic> json) {
    final bill = _map(json['bill']);
    return FetchedBill(
      fetchId: _text(json['fetchId']),
      service: _text(json['service']),
      providerId: _text(json['providerId']),
      providerName: _text(json['providerName']),
      accountRef: _text(json['accountRef']),
      rule: AmountRule.fromJson(_map(json['amountRule'])),
      customerName: _text(bill['customerName']),
      amount: _double(bill['amount']),
      dueDate: _text(bill['dueDate']),
      billDate: _text(bill['billDate']),
      billNumber: _text(bill['billNumber']),
      billPeriod: _text(bill['billPeriod']),
      billerName: _text(bill['billerName']),
      billerBalance: _text(bill['billerBalance']),
      remarks: _text(bill['remarks']),
      expiresAt: _date(json['expiresAt']),
    );
  }

  final String fetchId;
  final String service;
  final String providerId;
  final String providerName;
  final String accountRef;
  final AmountRule rule;
  final String customerName;
  final double amount;
  final String dueDate;
  final String billDate;
  final String billNumber;
  final String billPeriod;
  final String billerName;
  final String billerBalance;
  final String remarks;
  final DateTime? expiresAt;

  DateTime? get due => dueDate.isEmpty ? null : DateTime.tryParse(dueDate);

  bool get isExpired => expiresAt != null && DateTime.now().isAfter(expiresAt!);
}

class MobileOperatorMatch {
  const MobileOperatorMatch({
    required this.providerId,
    required this.providerName,
    this.circle = '',
    this.stateId = '',
  });

  factory MobileOperatorMatch.fromJson(Map<String, dynamic> json) => MobileOperatorMatch(
        providerId: _text(json['providerId']),
        providerName: _text(json['providerName']),
        circle: _text(json['circle']),
        stateId: _text(json['stateId']),
      );

  final String providerId;
  final String providerName;
  final String circle;
  final String stateId;
}

class RechargePlan {
  const RechargePlan({
    required this.type,
    required this.amount,
    this.detail = '',
    this.validity = '',
    this.talktime = '',
    this.data = '',
  });

  factory RechargePlan.fromJson(Map<String, dynamic> json) => RechargePlan(
        type: _text(json['type']).isEmpty ? 'OTHER' : _text(json['type']),
        amount: _double(json['amount']),
        detail: _text(json['detail']),
        validity: _text(json['validity']),
        talktime: _text(json['talktime']),
        data: _text(json['data']),
      );

  final String type;
  final double amount;
  final String detail;
  final String validity;
  final String talktime;
  final String data;
}

class RechargePlans {
  const RechargePlans({this.categories = const [], this.plans = const []});

  factory RechargePlans.fromJson(Map<String, dynamic> json) => RechargePlans(
        categories: (json['categories'] is List ? json['categories'] as List : const [])
            .map(_text)
            .where((c) => c.isNotEmpty)
            .toList(),
        plans: _maps(json['plans']).map(RechargePlan.fromJson).where((p) => p.amount > 0).toList(),
      );

  final List<String> categories;
  final List<RechargePlan> plans;
}

/// A recharge or bill payment being processed at the operator.
class ServiceTransaction {
  const ServiceTransaction({
    required this.urid,
    required this.status,
    this.type = '',
    this.service = '',
    this.amount = 0,
    this.providerName = '',
    this.accountRef = '',
    this.orderId = '',
    this.operatorTxnId = '',
    this.message = '',
    this.paymentId,
    this.refund,
    this.createdAt,
    this.completedAt,
  });

  factory ServiceTransaction.fromJson(Map<String, dynamic> json) {
    final refund = _map(json['refund']);
    return ServiceTransaction(
      urid: _text(json['urid']),
      status: _text(json['status']).isEmpty ? 'processing' : _text(json['status']).toLowerCase(),
      type: _text(json['type']),
      service: _text(json['service']),
      amount: _double(json['amount']),
      providerName: _text(json['providerName']),
      accountRef: _text(json['accountRef']),
      orderId: _text(json['orderId']),
      operatorTxnId: _text(json['operatorTxnId']),
      message: _text(json['message']),
      paymentId: _text(json['paymentId']).isEmpty ? null : _text(json['paymentId']),
      refund: refund.isEmpty ? null : refund,
      createdAt: _date(json['createdAt']),
      completedAt: _date(json['completedAt']),
    );
  }

  final String urid;
  final String status;
  final String type;
  final String service;
  final double amount;
  final String providerName;
  final String accountRef;
  final String orderId;
  final String operatorTxnId;
  final String message;
  final String? paymentId;
  final Map<String, dynamic>? refund;
  final DateTime? createdAt;
  final DateTime? completedAt;

  bool get isSuccess => status == 'completed';
  bool get isFailed => status == 'failed' || status == 'cancelled';
  bool get isPending => !isSuccess && !isFailed;

  String get refundStatus => _text(refund?['status']);

  /// Customer-facing refund line, or null when there is no refund.
  String? get refundLabel {
    switch (refundStatus) {
      case 'WALLET_CREDITED':
        return 'Refunded to your Khatu Pay wallet';
      case 'RAZORPAY_REFUNDED':
        return 'Refund initiated to your original payment method (5-7 working days)';
      case 'REVIEW_REQUIRED':
      case 'PROCESSING':
        return 'Refund is being processed';
      default:
        return null;
    }
  }
}
